field/TinyLang/Field/Typed/TypeChecker.hs

```diff
@@ -220,25 +220,32 @@ checkExpr m = do
     let uniMismatch = typeMismatch tM uni mUni
     withGeqUniM uni mUni uniMismatch tM
 
+-- TODO:  Figure how to optimise
 checkProgram
     :: forall m f. (MonadTypeChecker m, TextField f)
     => R.Program R.Var f -> m (T.Program f)
-checkProgram = traverse checkStatement
+checkProgram (R.Program (R.Statements stmts)) =
+    T.Program . T.Statements <$> foldMapA checkStatement stmts
 
 
 {-| Type checking judgement for statements of form
 -}
 checkStatement
     :: forall m f. (MonadTypeChecker m, TextField f)
-    => R.Statement R.Var f -> m (T.Statement f)
+    => R.Statement R.Var f -> m [T.Statement f]
 checkStatement (R.ELet var m) = do
     Some uniVar@(T.UniVar uni _) <- inferUniVar var
-    T.withKnownUni uni $ T.ELet uniVar <$> checkExpr m
+    T.withKnownUni uni $ (:[]) . T.ELet uniVar <$> checkExpr m
 checkStatement (R.EAssert m) =
-    T.EAssert <$> checkExpr m
-checkStatement (R.EFor var start end stmts) = do
-    tVar <- makeVar $ R.unVar var
-    T.EFor (T.UniVar Field tVar) start end <$> traverse checkStatement stmts
+    (:[]) . T.EAssert <$> checkExpr m
+-- TODO:  Inject var into environment of stmts
```

Comment:
Instead of having `Map String Var` as scope there and binding the loop variable via `let`, we could have `Map String (Either f Var)` and just inline the loop variable in the loop body. Not sure if it's a better strategy or not. The compiler will do that anyway.

Reply:
For nested loops where the inner loop references the loop variable of the outer loop you'll have to inline the value of the outer loop variable, so you'll have to track what loop variables map to anyway.

Reply:
Added a test for variable bindings in nested loops in fa382e0.

I think that constant folding and constant propagation should be done as separate steps that operate on `Typed.Core`.

Reply:
Sorry for commenting on the merged PR, I was a bit distracted and I'm now catching up. 

> I think that constant folding and constant propagation should be done as separate steps that operate on `Typed.Core`.

That's impossible, because in order to unroll

```
for i = 1 to 3 do
    for j = 1 to i do
       <...>
    end;
end;
```

you need to propagate `i` beforehand, so constant propagation should happen before (or at the same time as) loop unrolling.

Also, it'd be fine to unroll loops in the untyped language rather than doing that at the same time as type checking. But anyway, constant propagation should happen before (or at the same time as) loop unrolling.